DM-3510 CRA Client documentation (#3551)
SoftwareAG/c8y-docs: content/cloud-remote-access/cra-api-bundle/client-implementation.md

---

> title: Client Implementation

Reviewer: ```suggestion
title: Client implementation
```

weight: 10
layout: bundle
---
Each device supporting Cloud Remote Access uses a fragment called `c8y_RemoteAccessList` to hold a list with the configured
endpoints that can be accessed via this device.

```json
"c8y_RemoteAccessList": [
        {
            "serialVersionUID": 6652959747455810127,
            "hostname": "localhost",
            "protocol": "PASSTHROUGH",
            "credentials": {
                "privateKey": null,
                "password": null,
                "certificate": null,
                "publicKey": null,
                "hostKey": null,
                "type": "NONE",
                "username": null
            },
            "port": 33123,
            "name": "My HTTP Echo Server",
            "id": "1"
        }
    ]
```

> 

Reviewer 2: If we document this JSON above, we should also add a table explaining the properties, similar to other places.

Author: Done

In the example above a local HTTP echo server is reachable from the device at `http://localhost:33123`. The configuration id is `1`. 

> To connect a client application the client needs to open a websocket to the following URL:

Reviewer: ```suggestion
To connect a client application, the client must open a WebSocket to the following URL:
```

```http
wss://<tenant domain>/service/remoteaccess/client/<device id>/configurations/<configuration id>
```

> ### Required HTTP Headers

Reviewer: ```suggestion
### Required HTTP headers {required-http-headers}
```

> The websocket client needs to set the Websocket subprotocol to `binary` using the `Sec-WebSocket-Protocol` header.

Reviewer: ```suggestion
The WebSocket client must set the WebSocket subprotocol to `binary` using the `Sec-WebSocket-Protocol` header.
```

> The WebSocket client must set the WebSocket subprotocol to `binary` using the `Sec-WebSocket-Protocol` header.

Reviewer 2: This is technically not a requirement on the server side. This is header is evaluated when the server responds with the same header to the initial upgrade request. Our server will always respond with binary as protocol regardless of what the client has requested. If there is a mismatch between the requested protocol and the returned ones, most reputable websocket libraries will fail to establish the connection on the client side. So the binary protocol is technically only recommended.

Author: Fixed with latest commit

> In addition, a valid authorization header is required (see [OpenAPI:Authentication](https://cumulocity.com/api/core/#section/Authentication)).

Reviewer: ```suggestion
In addition, a valid authorization header is required, see [Authentication](https://cumulocity.com/api/core/#section/Authentication) in the {{< openapi >}}.
```

```http
Sec-WebSocket-Protocol: binary 
Authorization: <auth header>
```

> ### Sending and receiving traffic

Reviewer: ```suggestion
### Sending and receiving traffic {sending-and-receiving-traffic}
```

> Once the web socket connection is established, traffic for the endpoint can simply be sent to the websocket. Data from the endpoint can be consumed by reading from the websocket.

Reviewer: ```suggestion
Once the WebSocket connection is established, traffic for the endpoint can simply be sent to the WebSocket. Data from the endpoint can be consumed by reading from the WebSocket.
```

> To implement a local forwarding proxy, data from a local server socket needs to be simply written to the websocket and vice versa.

Reviewer: ```suggestion
To implement a local forwarding proxy, data from a local server socket must be written to the WebSocket and vice versa.
```

| `credentials.username`                 | String | See details | **Mandatory if `credentials.type` is `USER_PASS`, `KEY_PAIR`, or `CERTIFICATE`.** Otherwise, not applicable.                                                                                                                                                                                                                                         |

##### Example 

```json
"c8y_RemoteAccessList": [
        {
            "hostname": "localhost",
            "protocol": "PASSTHROUGH",
            "credentials": {
                "privateKey": null,
                "password": null,
                "certificate": null,
                "publicKey": null,
                "hostKey": null,
                "type": "NONE",
                "username": null
            },
            "port": 33123,
            "name": "My HTTP Echo Server",
            "id": "1"
        }
    ]
```

> In the example above a local HTTP echo server is reachable from the device at `http://localhost:33123`. The configuration id is `1`.

Reviewer: ```suggestion
In the example above a local HTTP echo server is reachable from the device at *http://localhost:33123*. The configuration ID is `1`.
```

To connect a client application, the client must open a WebSocket to the following URL.

```http
wss://<tenant domain>/service/remoteaccess/client/<device id>/configurations/<configuration id>
```

> ### Websocket HTTP headers {#websocket-http-headers}

Reviewer: ```suggestion
### WebSocket HTTP headers {#websocket-http-headers}
```

> The Cloud Remote Access service uses the `binary` websocket sub-protocol, regardless of which protocol is requested by the client.  Hence we recommend to pin the WebSocket subprotocol to `binary` using the `Sec-WebSocket-Protocol` header.

Reviewer: ```suggestion
The Cloud Remote Access service uses the binary WebSocket subprotocol, regardless of which protocol is requested by the client.  Hence, we recommend pinning the WebSocket subprotocol to `binary` using the `Sec-WebSocket-Protocol` header.
```
